### PR TITLE
bars dataset + priors

### PR DESCRIPTION
**Pull request by alvinzz**
Refactor BarsDataset.

Adds Priors:
   - Spike-and-Slab
   - L0-norm

Updates unittests.

### CODE REVIEW

**Comment by belsten** [comment]
General comment about formatting, we currently have a formatting style guide that is in another branch and will be merged. https://github.com/rctn/sparsecoding/blob/formatting/docs/contributing.md
Much of the stable code we have right now follows this guide.

**Comment by cshewmake2** [comment]
Thanks for all of the commits! For future reference: let's make sure each branch is adding its own feature. It's a lot easier to review and accept PRs quickly when they're `atomic` and add a specific new functionality or change.

**Comment by alvinz-matician** [comment]
Yeah old (bad) habits die hard. If it's too much I'm happy to break it up into smaller PRs. But it would be difficult to split everything from moving the BarsDataset to the last tests change, since the tests depend on that dataset.

File: sparsecoding/data/datasets/common.py [hunk 1]
```diff
@@ -0,0 +1 @@
+RAW_DATA_DIR = "raw_data/"
```
Comment (cshewmake2) [inline sparsecoding/data/datasets/common.py:1]: I like this idea; we should think about where to put this directory. I think this (combined with the OlshausenDataset class loading scheme) will make a new directory from wherever the user's present working directory is when it's called. 
I think this is what pytorch does with MNIST as well. 

Is there a better way to do this? Maybe this is fine.

**Comment by cshewmake2** [comment]
I have yet to test out everything on my local machine. Will try to do that tomorrow sometime.

File: sparsecoding/data/datasets/common.py (hunk 1 above)
Comment (alvinzz) [reply sparsecoding/data/datasets/common.py:1]: Oh yes, that's a good point. I will change to 
```
import sparsecoding
dir = sparsecoding.__file__
```
or something like that.

File: sparsecoding/priors/common.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
+from abc import ABC, abstractmethod
```
[the remark this answers is not in the document]
Comment (alvinzz) [reply sparsecoding/priors/common.py:1]: It doesn't really enforce anything (being Python and all), but semantically it's designed for this purpose:

https://peps.python.org/pep-3119/
```
ABCs are simply Python classes that are added into an object’s inheritance tree to signal certain features of that object to an external inspector. Tests are done using isinstance(), and the presence of a particular ABC means that the test has passed.

In addition, the ABCs define a minimal set of methods that establish the characteristic behavior of the type. Code that discriminates objects based on their ABC type can trust that those methods will always be present. Each of these methods are accompanied by an generalized abstract semantic definition that is described in the documentation for the ABC. These standard semantic definitions are not enforced, but are strongly recommended.
```

File: sparsecoding/data/datasets/common.py (hunk 1 above)
Comment (alvinzz) [reply sparsecoding/data/datasets/common.py]: done!

**Review by ligeralde — COMMENTED** [review]
make variable names all explicit (e.g., D --> dimension)

**Pull request merged**